### PR TITLE
Rebuilt ElasticSearch client on top of their REST API

### LINKED ISSUE #7471: Implement SkyWalking's native ElasticSearch protocol client for ElasticSearch 6/7/8 and OpenSearch 1.x
With the [SSPL impact](https://skywalking.apache.org/blog/2021-01-17-elastic-change-license/), we can't upgrade the HLRC anymore, unless Elastic eventually release APLv2 licensed client. But, even we have that, with current Elastic [current move](https://discuss.opendistrocommunity.dev/t/elastic-java-clients-are-no-longer-open-source/6470), the client is not usable for OpenSearch.
We(SkyWalking) don't choose side according to ASF vendor neutral guidance, but we have to follow License and laws. So, once SSPL client is there, we can't distribute with it.
Here is what I propose
1. We remove the ElasticSearch 6.x HLRC permanently, so, we don't need to release 2 binary distributions anymore.
2. Use HTTP Client(which lib could be determined later) + JSON for ES 6/7/8, which is also should work with OpenSearch.

For this new storage option, we could name it `sw-native-elasticsearch`.

### LINKED ISSUE #7655: [Bug] ElasticSearch storage `secretsManagementFile` hot reloading doesn't work as expected
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/skywalking/issues?q=is%3Aissue) and found no similar issues.

### Apache SkyWalking Component

OAP server (apache/skywalking)

### What happened

The `BulkProcessor` created here is bound to the `client` and the `BulkProcessor` is not updated afterwards.

https://github.com/apache/skywalking/blob/de59bf7aebc20110c21c0e39e9b987e4aec91911/oap-server/server-library/library-client/src/main/java/org/apache/skywalking/oap/server/library/client/elasticsearch/ElasticSearchClient.java#L569-L578

When the `secretsManagementFile` is updated, the `client` will be re-created in `elasticSearchClient.connect()`, and the original `BulkProcessor` is still bound to the old ElasticSearch client, which should not work.

https://github.com/apache/skywalking/blob/69ddc07cc152c415f385f863baff65c9c6079e2b/oap-server/server-storage-plugin/storage-elasticsearch-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/elasticsearch/StorageModuleElasticsearchProvider.java#L145-L148

### What you expected to happen

The `BulkProcessor` should use the newly-created ElasticSearch client, or the ElasticSearch client is not created, only update the needed fields.

### How to reproduce

I notice this when reviewing the codes in #7634 , but haven't fixed it in #7634 , we'd better fix this in 8.8.0

### Anything else

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by kezhenxu94**
- [x] If this pull request closes/resolves/fixes an existing issue, replace the issue number. Closes #7471 , closes #7655.
- [x] Update the [`CHANGES` log](https://github.com/apache/skywalking/blob/master/CHANGES.md).

### CODE REVIEW

**Comment by kezhenxu94** [comment]
Mark it as ready for review for reviewers who want to jump in as earlier as possible, though there are a lot "TODO" and to be optimized

File: oap-server/server-library/library-elasticsearch-client/pom.xml
Comment (wu-sheng) [inline oap-server/server-library/library-elasticsearch-client/pom.xml:35]: Good to have this. FYI @apache/skywalking-committers

File: oap-server/server-library/library-elasticsearch-client/src/main/java/org/apache/skywalking/library/elasticsearch/ElasticSearchVersion.java
Comment (wu-sheng) [inline oap-server/server-library/library-elasticsearch-client/src/main/java/org/apache/skywalking/library/elasticsearch/ElasticSearchVersion.java:59]: Are We going to test v8 in this PR?

File: oap-server/server-library/library-elasticsearch-client/src/main/java/org/apache/skywalking/library/elasticsearch/bulk/BulkProcessor.java
Comment (wu-sheng) [inline oap-server/server-library/library-elasticsearch-client/src/main/java/org/apache/skywalking/library/elasticsearch/bulk/BulkProcessor.java:63]: One thread only? It seems `concurrentRequests` not being used.

File: oap-server/server-library/library-elasticsearch-client/src/main/java/org/apache/skywalking/library/elasticsearch/ElasticSearchVersion.java
Comment (kezhenxu94) [reply oap-server/server-library/library-elasticsearch-client/src/main/java/org/apache/skywalking/library/elasticsearch/ElasticSearchVersion.java:59]: > Are We going to test v8 in this PR?

No, this PR is already large enough

File: oap-server/server-starter/src/main/java/org/apache/skywalking/oap/server/starter/OAPServerStartUp.java
Comment (wu-sheng) [inline oap-server/server-starter/src/main/java/org/apache/skywalking/oap/server/starter/OAPServerStartUp.java:21]: Considering ES7 starter removed, do you still need to keep this module? How about moving this into bootstrap directly?

File: docs/en/setup/backend/backend-storage.md
Comment (wu-sheng) [inline docs/en/setup/backend/backend-storage.md:48]: I think we still need to indicate the supported-versions range.

File: oap-server/server-configuration/configuration-apollo/pom.xml
Comment (wu-sheng) [inline oap-server/server-configuration/configuration-apollo/pom.xml:61]: Why do we need to add a test dependency but no code(test code) change?

Comment (kezhenxu94) [reply oap-server/server-configuration/configuration-apollo/pom.xml:61]: > Why do we need to add a test dependency but no code(test code) change?

This dependency was brought by ElasticSearch HLRC in `library-client` (as runtime dependency), since ES HLRC is removed, we have to add it explicitly, the same for other modules dependencies.

**Comment by wu-sheng** [comment]
I haven't reviewed the client's implementations, as you mentioned `TODO` should be fixed later. But besides that, this is generally good to me.

File: docs/en/setup/backend/backend-storage.md
Comment (kezhenxu94) [reply docs/en/setup/backend/backend-storage.md:48]: Added

File: oap-server/server-library/library-elasticsearch-client/src/main/java/org/apache/skywalking/library/elasticsearch/ElasticSearch.java
Comment (wu-sheng) [inline oap-server/server-library/library-elasticsearch-client/src/main/java/org/apache/skywalking/library/elasticsearch/ElasticSearch.java:85]: How should we set up thread number of the web client(or netty)? I think this matters in query or metrics(including metadata) update?

Comment (kezhenxu94) [reply oap-server/server-library/library-elasticsearch-client/src/main/java/org/apache/skywalking/library/elasticsearch/ElasticSearch.java]: Added configuration

File: oap-server/server-library/library-elasticsearch-client/src/main/java/org/apache/skywalking/library/elasticsearch/bulk/BulkProcessor.java
Comment (kezhenxu94) [reply oap-server/server-library/library-elasticsearch-client/src/main/java/org/apache/skywalking/library/elasticsearch/bulk/BulkProcessor.java]: Added

**Comment by kezhenxu94** [comment]
Now the PR should be good to go

File: docs/en/setup/backend/configuration-vocabulary.md
Comment (wu-sheng) [inline docs/en/setup/backend/configuration-vocabulary.md:88]: Is this the default of official ElasticSearch client?  Should this be 2 or number of core aware?

Comment (kezhenxu94) [reply docs/en/setup/backend/configuration-vocabulary.md]: > Is this the default of official ElasticSearch client? Should this be 2 or number of core aware?

The default of ES client is num of processors, updated

File: docs/en/setup/backend/configuration-vocabulary.md
Comment (wu-sheng) [inline docs/en/setup/backend/configuration-vocabulary.md:88]: Just one check, all metrics/log/traces persistent are according to bulk thread today, right? So, it always uses 2 threads in default. Then others are for query if there are over 2 cores.

**Review by wu-sheng — APPROVED** [review]
LGTM. Wait for others' review.

**Comment by JaredTan95** [comment]
I tested locally, all looks great to me except `EmptyEndpointGroupException` as follows when `es` is shutdown and connected failures:
```bash
2021-09-06 11:14:10,020 org.apache.skywalking.library.elasticsearch.ElasticSearch 149 [armeria-eventloop-nio-4-2] DEBUG [] - [creqId=3fe1891d, preqId=8179d41f][http://UNKNOWN/#GET] Request: {startTime=2021-09-06T03:14:10.010Z(1630898050010987), length=0B, duration=4538µs(4538520ns), cause=com.linecorp.armeria.client.UnprocessedRequestException: com.linecorp.armeria.client.endpoint.EmptyEndpointGroupException, scheme=none+http, name=GET, headers=[:method=GET, :path=/, :scheme=http, :authority=UNKNOWN]}
2021-09-06 11:14:10,021 org.apache.skywalking.library.elasticsearch.ElasticSearch 186 [armeria-eventloop-nio-4-2] WARN  [] - [creqId=3fe1891d, preqId=8179d41f][http://UNKNOWN/#GET] Response: {startTime=2021-09-06T03:14:10.020Z(1630898050020718), length=0B, duration=0ns, totalDuration=9749µs(9749393ns), cause=com.linecorp.armeria.client.UnprocessedRequestException: com.linecorp.armeria.client.endpoint.EmptyEndpointGroupException, headers=[:status=0]}
com.linecorp.armeria.client.UnprocessedRequestException: com.linecorp.armeria.client.endpoint.EmptyEndpointGroupException
```
I think `EmptyEndpointGroupException` is not enough clear to point out the cause of the error?

**Comment by kezhenxu94** [comment]
> I tested locally, all looks great to me except `EmptyEndpointGroupException` as follows when `es` is shutdown and connected failures:
> 
> ```shell
> 2021-09-06 11:14:10,020 org.apache.skywalking.library.elasticsearch.ElasticSearch 149 [armeria-eventloop-nio-4-2] DEBUG [] - [creqId=3fe1891d, preqId=8179d41f][http://UNKNOWN/#GET] Request: {startTime=2021-09-06T03:14:10.010Z(1630898050010987), length=0B, duration=4538µs(4538520ns), cause=com.linecorp.armeria.client.UnprocessedRequestException: com.linecorp.armeria.client.endpoint.EmptyEndpointGroupException, scheme=none+http, name=GET, headers=[:method=GET, :path=/, :scheme=http, :authority=UNKNOWN]}
> 2021-09-06 11:14:10,021 org.apache.skywalking.library.elasticsearch.ElasticSearch 186 [armeria-eventloop-nio-4-2] WARN  [] - [creqId=3fe1891d, preqId=8179d41f][http://UNKNOWN/#GET] Response: {startTime=2021-09-06T03:14:10.020Z(1630898050020718), length=0B, duration=0ns, totalDuration=9749µs(9749393ns), cause=com.linecorp.armeria.client.UnprocessedRequestException: com.linecorp.armeria.client.endpoint.EmptyEndpointGroupException, headers=[:status=0]}
> com.linecorp.armeria.client.UnprocessedRequestException: com.linecorp.armeria.client.endpoint.EmptyEndpointGroupException
> ```
> 
> I think `EmptyEndpointGroupException` is not enough clear to point out the cause of the error?

There is warning / error message before what you posted, also, it has health check registered .

```
2021-09-06 12:52:21,329 org.apache.skywalking.oap.server.telemetry.api.HealthCheckMetrics 50 [armeria-common-worker-nio-2-1] WARN  [] - Health check fails. reason: No healthy endpoint
2021-09-06 12:52:21,330 org.apache.skywalking.oap.server.library.client.elasticsearch.ElasticSearchClient 146 [armeria-common-worker-nio-2-1] ERROR [] - Failed to recreate ElasticSearch client based on new config
```

**Review by JaredTan95 — DISMISSED** [review]
LGTM

**Review by wu-sheng — APPROVED** [review]

**Comment by kezhenxu94** [comment]
Thank you @wu-sheng @JaredTan95 for reviewing and checking, I'm going to merge this

**Comment by wu-sheng** [comment]
@wankai123 is confirming too.

**Review by hanahmily — COMMENTED** [review]
`docker/oap/Dockerfile.oap` will be fixed in follow-up PRs?

**Comment by hanahmily** [comment]
> `docker/oap/Dockerfile.oap` will be fixed in follow-up PRs?

ignore this

**Comment by kezhenxu94** [comment]
> `docker/oap/Dockerfile.oap` will be fixed in follow-up PRs?

I think all needed changes to docker (and `docker/oap/Dockerfile.oap`) are included in this PR, the official docker images will be done in http://github.com/apache/skywalking-docker after this PR, if I missed anything, please let me know @hanahmily

**Review by hanahmily — COMMENTED** [review]
Some nits:

 * remove https://github.com/apache/skywalking/blob/master/docker/.env#L1
 * update https://github.com/apache/skywalking/blob/master/docker/.env#L2 to ES7

**Review by hanahmily — DISMISSED** [review]
LGTM

some nits regarding `docker/.env`

**Review by wankai123 — DISMISSED** [review]
LGTM

**Review by mrproliu — APPROVED** [review]

**Pull request merged**
